### PR TITLE
Subsidy

### LINKED ISSUE #66: [frontend] Show user balance and spend from subsidy

### LINKED ISSUE #67: [circuits] Allow a user to prove negative reputation

### LINKED ISSUE #68: [contracts] Allow negative rep proof to collect airdrop from subsidy

### LINKED ISSUE #69: [contracts] Allow action to spend from subsidy or rep proof

### LINKED ISSUE #76: Reputation creation
Right now we give people 30 rep every epoch. Instead we could use a system where users get a 30 rep subsidy each epoch. The first 30 rep spent is covered by the system. After that it's taken out of the users rep balance. If the user doesn't spend or receive any rep in an epoch their balance is unchanged.

- rep is only created by interactions in the system
- people can't hoard rep by getting airdrops and never spending
- if people create posts/comments just for exchanging rep others can vote and take rep away (vote negatively on the posts/comment for free)

This would also make the case of missing an epoch more clean. Right now if a user doesn't do a UST for an epoch they miss 30 rep for each epoch they skip.

### PR DESCRIPTION
**Pull request by vimwitch**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Depends #30 

Closes #66
Closes #67
Closes #68
Closes #69 
Closes #76 

## What is the new behavior?

In this version of the subsidy system there is a separate subsidy key proof that is used to spend.

- Subsidy spends cannot be combined with regular spends
- Subsidy spends always come from first epoch key
- Subsidy spends cannot be used to vote for yourself (must prove receiving key non-ownership)

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

**Comment by vimwitch** [comment]
This needs some more tests, especially for negative rep proofs in the contracts, but should mostly be ready for review. Demo running [here](http://test.unirep.io:30001/). Made a new task for building the negative rep proof into the ui #75

**Comment by kittybest** [comment]
- user cannot vote themselves (need UI hint)
- go to etherscan error: originally it's `kovan-optimistic.etherscan.io/tx/${post.id}`, but I found that the `post.id` is not transaction hash now? (instead, it's `data._id` now, according to `utils.ts` in frontend
- progress bar display has wrong number such as `108%`
- logout still not work
- if the user sign out and re-login after epoch transition, won't do UST automatically. (but if reload page, will do UST)
- user page is still full of bugs (1. not showing correct spent 2. not showing posts and comments with old epoch keys 3. click `activities` cause error)

The bugs can be fixed in other PR, for the core function of subsidy (use subsidy to post/comment/vote others, receive voted points after UST) is fine now :)

**Review by kittybest — APPROVED** [review]

File: packages/circuits/circuits/proveNegativeReputation.circom [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+include "../node_modules/circomlib/circuits/comparators.circom";
+include "../node_modules/circomlib/circuits/gates.circom";
+include "../node_modules/circomlib/circuits/mux1.circom";
+include "../node_modules/circomlib/circuits/poseidon.circom";
+include "../node_modules/circomlib/circuits/bitify.circom";
+include "../node_modules/circomlib/circuits/sign.circom";
+include "./sparseMerkleTree.circom";
+include "./identityCommitment.circom";
+include "./incrementalMerkleTree.circom";
+include "./modulo.circom";
```
Comment (vivianjeng) [inline packages/circuits/circuits/proveNegativeReputation.circom:17]: I think these files can be imported from `../node_modules/@unirep/circuits/circuits/***.circom`.
So we don't need the copy of `identityCommitment.circom`, ... and so on

File: packages/circuits/circuits/proveSubsidyKey.circom [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+include "../node_modules/circomlib/circuits/comparators.circom";
+include "../node_modules/circomlib/circuits/gates.circom";
+include "../node_modules/circomlib/circuits/mux1.circom";
+include "../node_modules/circomlib/circuits/poseidon.circom";
+include "../node_modules/circomlib/circuits/bitify.circom";
+include "../node_modules/circomlib/circuits/sign.circom";
+include "./sparseMerkleTree.circom";
+include "./identityCommitment.circom";
+include "./incrementalMerkleTree.circom";
+include "./modulo.circom";
```
Comment (vivianjeng) [inline packages/circuits/circuits/proveSubsidyKey.circom:17]: they can also be imported from `@unirep/circuits`

File: packages/circuits/src/Reputation.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
+import assert from 'assert'
+import * as crypto from '@unirep/crypto'
+
+export class Reputation {
```
Comment (vivianjeng) [inline packages/circuits/src/Reputation.ts:4]: I think this class can be imported from `@unirep/core`

File: packages/circuits/test/proveNegativeReputation.test.ts [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+
+const verifyProof = async (publicSignals, proof): Promise<boolean> => {
+    const vkey = require(path.join(
+        __dirname,
+        `../zksnarkBuild/${circuitName}.vkey.json`
+    ))
+    return snarkjs.groth16.verify(vkey, publicSignals, proof)
+}
+
+const genEpochKey = (
```
Comment (vivianjeng) [inline packages/circuits/test/proveNegativeReputation.test.ts:99]: it can be imported from `@unirep/core`

File: packages/circuits/test/proveSubsidyKey.test.ts [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+} from '@unirep/circuits'
+import { Reputation } from '../src/Reputation'
+
+const circuitName = 'proveSubsidyKey'
+const circuitPath = path.join(
+    __dirname,
+    `../circuits/test/${circuitName}_test.circom`
+)
+
+const genEpochKey = (
```
Comment (vivianjeng) [inline packages/circuits/test/proveSubsidyKey.test.ts:21]: it can be imported from `@unirep/core` as well

File: packages/circuits/circuits/proveSubsidyKey.circom [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    signal output subsidyKey;
+    signal output epochKey;
+
+    // we only need to verify that one epk is in the gst
+    // we can then simply calculate the others
+    /* 1. Calculate subsidy key and epoch key */
+
+    component subsidyKeyHasher = Poseidon(2);
+
+    subsidyKeyHasher.inputs[0] <== identity_nullifier + NUM_EPOCH_KEY_NONCE_PER_EPOCH;
```
Comment (vivianjeng) [inline packages/circuits/circuits/proveSubsidyKey.circom:49]: Attestations sent to the `subsidyKey` will not be processed in the unirep protocol
It will not be calculated in the [`processAttestations.circom`](https://github.com/Unirep/Unirep/blob/main/packages/circuits/circuits/processAttestations.circom#L84) and [`userStateTransition.circom`](https://github.com/Unirep/Unirep/blob/a9e808a3435fc14ff9aebb9d2dbb3147168b6b35/packages/circuits/circuits/userStateTransition.circom#L123)

The protocol will process nonce from `0` to `NUM_EPOCH_KEY_NONCE_PER_EPOCH-1`

File: packages/circuits/test/proveSubsidyKey.test.ts [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+    ))
+    return snarkjs.groth16.verify(vkey, publicSignals, proof)
+}
+
+const calcSubsidyKey = (id, epoch) => {
+    return crypto.hash2([
+        BigInt(id.identityNullifier) + BigInt(NUM_EPOCH_KEY_NONCE_PER_EPOCH),
+        BigInt(epoch),
+    ])
+}
```
Comment (vivianjeng) [inline packages/circuits/test/proveSubsidyKey.test.ts:121]: it should be able to use a `genEpochKey` to generate epoch key
e.g.
```typescript
const calcSubsidyKey = (id, epoch) => {
    const nonce = NUM_EPOCH_KEY_NONCE_PER_EPOCH
    return genEpochKey(id, epoch, nonce)
}
```

But the user will never receive reputation with this epoch key because the nonce cannot be equal to `NUM_EPOCH_KEY_NONCE_PER_EPOCH`

File: packages/core/contracts/NegativeRepVerifier.sol [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+        p1[2] = c1;
+        p1[3] = d1;
+        p2[0] = a2;
+        p2[1] = b2;
+        p2[2] = c2;
+        p2[3] = d2;
+        return pairing(p1, p2);
+    }
+}
+contract Verifier {
```
Comment (vivianjeng) [inline packages/core/contracts/NegativeRepVerifier.sol:164]: I think the Verifiers can be ignore to upload

File: packages/core/test/Subsidy.ts [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+import { ethers } from 'hardhat'
+import { expect } from 'chai'
+import { ZkIdentity } from '@unirep/crypto'
+import { genEpochKey } from '@unirep/core'
+import { genUserState } from './utils'
+import { deployUnirep } from '@unirep/contracts/deploy'
+import { deployUnirepSocial } from '../src/utils'
+import * as config from '@unirep/circuits'
+
+describe('Subsidy', function () {
```
Comment (vivianjeng) [inline packages/core/test/Subsidy.ts:11]: Should add tests to see after user state transition
if user can receive reputation with the subsidy epoch key.
and after epoch transition user can request more subsidy for the next epoch

Comment (vivianjeng) [reply packages/core/test/Subsidy.ts:11]: and neg reputation users should successfully get airdrop and come back to the system

File: packages/core/contracts/UnirepSocial.sol [hunk 10]
```diff
@@ -6,11 +6,29 @@ import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
 import "@openzeppelin/contracts/utils/math/SafeMath.sol";
 import "@openzeppelin/contracts/utils/Address.sol";
 import { Unirep } from "@unirep/contracts/Unirep.sol";
+import { zkSNARKHelper } from '@unirep/contracts/libraries/zkSNARKHelper.sol';
+
+interface IVerifier {
```
Comment (vivianjeng) [inline packages/core/contracts/UnirepSocial.sol:11]: It can be imported from `@unirep/contracts/interfaces/IVerifier.sol`

File: packages/core/contracts/UnirepSocial.sol [hunk 11]
```diff
@@ -6,11 +6,29 @@ import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
...
+        uint[2][2] memory b,
+        uint[2] memory c,
+        uint[7] memory input
+    ) external view returns (bool r);
+    function verifyProof(
+        uint[2] memory a,
+        uint[2][2] memory b,
+        uint[2] memory c,
+        uint[6] memory input
+    ) external view returns (bool r);
```
Comment (vivianjeng) [inline packages/core/contracts/UnirepSocial.sol:23]: duplicated?

**Review by vivianjeng — CHANGES_REQUESTED** [review]
Please check if user uses subsidy, they can successfully receive votes after user state transition

File: packages/circuits/circuits/proveSubsidyKey.circom (hunk 6 above)
Comment (vimwitch) [reply packages/circuits/circuits/proveSubsidyKey.circom:49]: Yes, the subsidyKey is just used for tracking the spent subsidy. The circuit outputs epoch key with nonce 0. I think I could actually remove the subsidy key and just use epoch key 0.

File: packages/core/contracts/UnirepSocial.sol [hunk 12]
```diff
@@ -101,6 +130,167 @@ contract UnirepSocial {
...
+        // Submit attestation to receiver's first epoch key
+        Unirep.Attestation memory attestation;
+        attestation.attesterId = attesterId;
+        attestation.posRep = receivedSubsidy;
+        attestation.negRep = 0;
+        unirep.submitAttestation{value: attestingFee}(
+            attestation,
+            publicSignals[2] // first epoch key
+        );
+    }
```
Comment (vivianjeng) [inline packages/core/contracts/UnirepSocial.sol:205]: should prevent users from requesting subsidy airdrop more than twice.
Because user can always generate negative rep proof (e.g. `-3`) in the same epoch
so if user request for 10 times with `-3` proof, the user can get 30 reputation after user state transition

File: packages/circuits/circuits/proveNegativeReputation.circom (hunk 1 above)
Comment (vimwitch) [reply packages/circuits/circuits/proveNegativeReputation.circom:17]: This causes path issues. Our unirep circuits files require iden3 files from `../../node_modules` which breaks.

File: packages/circuits/src/Reputation.ts (hunk 3 above)
Comment (vimwitch) [reply packages/circuits/src/Reputation.ts:4]: Yes, it's not currently exported so I duplicated it

File: packages/core/contracts/UnirepSocial.sol (hunk 10 above)
Comment (vimwitch) [reply packages/core/contracts/UnirepSocial.sol:11]: I'm using the circom verifier generator so the interface is different

File: packages/core/contracts/UnirepSocial.sol (hunk 11 above)
Comment (vimwitch) [reply packages/core/contracts/UnirepSocial.sol:23]: Different number of public signals (fourth argument).

File: packages/core/contracts/UnirepSocial.sol (hunk 12 above)
Comment (vimwitch) [reply packages/core/contracts/UnirepSocial.sol:205]: 7d6ff15c9348195821478246342dc9af8ca3cd43

Comment (vimwitch) [reply packages/core/contracts/UnirepSocial.sol]: 7d6ff15c9348195821478246342dc9af8ca3cd43

File: packages/circuits/test/proveNegativeReputation.test.ts (hunk 4 above)
Comment (vimwitch) [reply packages/circuits/test/proveNegativeReputation.test.ts]: d9b016e30202f701fb47e0e7db4efdfa0e79bb9f

File: packages/circuits/test/proveSubsidyKey.test.ts (hunk 5 above)
Comment (vimwitch) [reply packages/circuits/test/proveSubsidyKey.test.ts]: d9b016e30202f701fb47e0e7db4efdfa0e79bb9f

File: packages/core/test/Subsidy.ts (hunk 9 above)
Comment (vimwitch) [reply packages/core/test/Subsidy.ts:11]: e09f9de74f79279a8bc77ca10dbb18bdbb57b820

File: packages/core/contracts/NegativeRepVerifier.sol (hunk 8 above)
Comment (vimwitch) [reply packages/core/contracts/NegativeRepVerifier.sol]: 8d4e351bc4a3312f846f85764a7aa21bb958cb56

File: packages/circuits/circuits/proveSubsidyKey.circom [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+    GST_root <== merkletree.root;
+
+    /* End of check 2 */
+
+    /* 3. Check if the claimed reputation given by the attester is in the user state tree */
+    component reputation_hasher = Poseidon(5);
+    reputation_hasher.inputs[0] <== pos_rep;
+    reputation_hasher.inputs[1] <== neg_rep;
+    reputation_hasher.inputs[2] <== graffiti;
+    reputation_hasher.inputs[3] <== sign_up;
```
Comment (vivianjeng) [inline packages/circuits/circuits/proveSubsidyKey.circom:79]: The same here.
we should check `sign_up === 1`

File: packages/frontend/src/context/User.ts [hunk 14]
```diff
@@ -433,27 +480,27 @@ export class User {
...
+    //         BigInt(decodedData.attestation.signUp)
+    //     )
+    //     const normalizedEpk = epochKey
+    //         .toHexString()
+    //         .replace('0x', '')
+    //         .padStart(this.unirepConfig.epochTreeDepth / 4, '0')
+    //     if (this.currentEpochKeys.indexOf(normalizedEpk) !== -1) {
+    //         this.spent += Number(attestation.negRep)
+    //     }
+    // }
```
Comment (vivianjeng) [inline packages/frontend/src/context/User.ts:503]: if we don't need this we can remove here.

File: packages/circuits/circuits/proveNegativeReputation.circom [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+    /* End of check 2 */
+
+
+    /* 3. Check if the claimed reputation given by the attester is in the user state tree */
+    component reputation_hasher = Poseidon(5);
+    reputation_hasher.inputs[0] <== pos_rep;
+    reputation_hasher.inputs[1] <== neg_rep;
+    reputation_hasher.inputs[2] <== graffiti;
+    reputation_hasher.inputs[3] <== sign_up;
```
[the remark this answers is not in the document]
Comment (vivianjeng) [reply packages/circuits/circuits/proveNegativeReputation.circom:67]: It would be better to add tests for this.

**Review by vivianjeng — CHANGES_REQUESTED** [review]

**Review by vivianjeng — APPROVED** [review]
LGTM
